CharlieBackend.Business/Services/DashboardService.cs

```diff
@@ -0,0 +1,288 @@
+                    .GetStudentsPresenceListByStudentIds(studentsIds);
+            }
+
+            if (request.IncludeAnalytics.Contains(ClassbookResultType.StudentMarks))
+            {
+                result.StudentsMarks = await _dashboardRepository
+                    .GetStudentsMarksListByStudentIds(studentsIds);
+            }
+
+            return Result<StudentsClassbookResultDto>.GetSuccess(result);
+        }
+
+        public async Task<Result<StudentsResultsDto>> GetStudentsResultAsync(StudentsRequestDto<StudentResultType> request)
+        {
+            if (ValidateStudentsRequest(request).Any())
+            {
+                return Result<StudentsResultsDto>.GetError(ErrorCode.ValidationError,
+                    String.Join(";\n", ValidateStudentsRequest(request)));
+            }
+
+            var result = new StudentsResultsDto();
+            var studentsIds = new List<long>();
+
+            if (request.StudentGroupId != default)
+            {
+                studentsIds = await _dashboardRepository.GetStudentsIdsByGroupIdsAsync(
+                    new List<long> { (long)request.StudentGroupId });
+            }
+            else
+            {
+                var studentGroupsIds = await _dashboardRepository
+                    .GetGroupsIdsByCourseIdAsync((long)request.CourseId, request.StartDate, request.FinishDate);
+
+                studentsIds = await _dashboardRepository
+                    .GetStudentsIdsByGroupIdsAsync(studentGroupsIds);
+            }
+
+            if (request.IncludeAnalytics.Contains(StudentResultType.AverageStudentMark))
+            {
+                result.AverageStudentsMarks = await _dashboardRepository
+                    .GetStudentsAverageMarksByStudentIdsAsync(studentsIds);
+            }
+
+            if (request.IncludeAnalytics.Contains(StudentResultType.AverageStudentVisits))
+            {
+                result.AverageStudentVisits = await _dashboardRepository
+                    .GetStudentsAverageVisitsByStudentIdsAsync(studentsIds);
+            }
+
+            return Result<StudentsResultsDto>.GetSuccess(result);
+        }
+
+        public async Task<Result<StudentsClassbookResultDto>> GetStudentClassbookAsync(long studentId,
+            GenericRequestDto<ClassbookResultType> request, ClaimsPrincipal userContext)
+        {
+            if (ValidateGenericRequest(request).Any())
+            {
+                return Result<StudentsClassbookResultDto>
+                           .GetError(ErrorCode.ValidationError, String.Join(";\n",
+                           ValidateGenericRequest(request)));
+            }
+            else if (ValidateStudentRights(studentId, userContext).Any())
+            {
+                return Result<StudentsClassbookResultDto>
+                            .GetError(ErrorCode.Unauthorized, String.Join(";\n",
+                            ValidateStudentRights(studentId, userContext)));
+            }
```

Comment:
1. code duplicate
2. you're calling validation methods two times - call them once, merge in single list and then make one check

Reply:
I did to show to front correct CodeError in every issue type. Ok, will fix